daemon/run.go

```diff
@@ -0,0 +1,319 @@
+func getServiceID(requestVars map[string]string) (akid.ServiceID, *HTTPResponse) {
+	var httpErr *HTTPResponse = nil
+
+	serviceName := requestVars["serviceName"]
+	frontClient := rest.NewFrontClient(cmdArgs.Domain, cmdArgs.ClientID)
+	result, err := util.GetServiceIDByName(frontClient, serviceName)
+	if err != nil {
+		httpErr = NewHTTPError(err, http.StatusNotFound, "Service not found")
+	}
+	return result, httpErr
+}
+
+// Creates a new learning session in the Akita back end.
+func createLearnSession(request *http.Request) HTTPResponse {
+	vars := mux.Vars(request)
+
+	// Get the service ID.
+	serviceID, httpErr := getServiceID(vars)
+	if httpErr != nil {
+		return *httpErr
+	}
+
+	// Create a new learning session with a random name.
+	// TODO: Allow user to specify a name?
+	learnSessionName := util.RandomLearnSessionName()
+	tags := map[string]string{}
+	_, err := util.NewLearnSession(cmdArgs.Domain, cmdArgs.ClientID, serviceID, learnSessionName, tags, nil)
+	if err != nil {
+		return *NewHTTPError(err, http.StatusInternalServerError, "Unable to start learning session")
+	}
+
+	// Return an HTTPResponse with the name of the new learning session.
+	// XXX Return session ID instead? Return both name and ID? Will either cause issues when used as part of URI?
+	return *NewHTTPResponse(http.StatusOK,
+		struct {
+			LearnSessionName string `json:"learnSessionName"`
+		}{
+			LearnSessionName: learnSessionName,
+		})
+}
+
+// Adds a set of witnesses to a learning session in the Akita back end.
+//
+// The request payload is expected to contain a sequence of HAR entries in JSON-serialized format. This entry is added as a witness to the learning session identified in the request.
+func addWitnesses(request *http.Request) HTTPResponse {
+	vars := mux.Vars(request)
+
+	// Ensure the request body is JSON-encoded.
+	if httpErr := EnsureJSONEncodedRequestBody(request); httpErr != nil {
+		return *httpErr
+	}
+
+	// Get the service ID.
+	serviceID, httpErr := getServiceID(vars)
+	if httpErr != nil {
+		return *httpErr
+	}
+
+	// Get the learning session ID.
+	learnSessionName := vars["learnSessionName"]
+	learnClient := rest.NewLearnClient(cmdArgs.Domain, cmdArgs.ClientID, serviceID)
+	learnSessionID, err := util.GetLearnSessionIDByName(learnClient, learnSessionName)
+	if err != nil {
+		return *NewHTTPError(err, http.StatusNotFound, "Learning session not found")
+	}
+
+	successfulEntries := 0
+	sampledErrs := sampled_err.Errors{SampleCount: 3}
+
+	// Start a consumer goroutine that will read witnesses from a channel and pass them on to a backend collector.
+	witnessChannel := make(chan *Witness, WITNESS_BUFFER_SIZE)
+	doneChannel := make(chan bool)
+	go func() {
+		defer func() { doneChannel <- true }()
+
+		// Create collector for ingesting the witnesses.
+		collector := trace.NewBackendCollector(serviceID, learnSessionID, learnClient, kgxapi.Inbound, cmdArgs.Plugins)
+
+		// TODO: Offer option to remove third-party trackers?
```

Comment:
Yes, this would be great.  I believe there's logic to do this already in the CLI.  Can we factor it out so it's uniformly applied?

Not necessary for this PR, but necessary before we release this.

Reply:
Indeed, the logic is there. The question is, how do we want it applied? Per request, or per daemon instance (i.e., specified in each request or on the command line)?

Reply:
On learn session creation, maybe?  That gives the client the option to change things without redeploying (e.g. by changing config options in the environment), but it doesn't add an argument that's unlikely to change to every witness request.

Reply:
This would introduce state that we currently don't keep. The dæmon itself would need to remember whether tracker removal is desired, and this state would need to persist across restarts of the dæmon. My vote is for the flag to be provided at the start of each `addWitnesses` call. I'm expecting there to be multiple witnesses for each call, which would amortize the overhead.

Reply:
Sounds good to me.

Reply:
Done in b2ec1c3.